pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+                bracketing = False
+                bisect_or_secant = True
+                bi_sec_generator = self.__bisect_or_secant(0.0, c[j])
+                while bisect_or_secant is True:
+                    obj_and_grad, output = next(bi_sec_generator)
+                    if obj_and_grad is True:
+                        yield True, output
+                    else:
+                        # finished __bisect_or_secant
+                        bisect_or_secant = False
+                        # returning new brackets
+                        # and ending __initial_bracket()
+                        yield False, output
+
+            # The proposed point obeys the opposite slope condition
+            # at the lower bound
+            # i.e. dfdx_d < 0.0 and f_d <= self._current_f + self.epsilon.
+            # Checking this is unnecessary as it is the opposite to
+            # the above conditions. This means the bracket interval needs
+            # expanding to ensure a minimum is bracketed.
+            # (steps B3 from [1])
+            else:
+                # Increasing the proposed point by a factor of row to attempt
+                # to bracket minimum and trying again.
+                c.append(rho * c[j])
+
+                # Increamenting the iteration counter.
+                j += 1
+
+    def __update(self, a, b, c):
+        '''
+        This function is part of the Hager-Zhang line search method [1].
+
+        Updates the bracketing boundary values of alpha. Ensuring the opposite
+        slope conditions are obeyed.
+
+        The opposite slope conditions are:
+                                φ(a) ≤ φ(0) + epsilon,
+                                φ'(a) < 0 (w.r.t every parameter),
+                                φ'(b) ≥ 0 (w.r.t every parameter).
+        where φ(α) = f(x+α*px), f() is the function binging minimised, x is
+        the current mparameter set, px is the newton direction, and alpha is
+        the step size.
+
+        In the first condition, epsilon is a small positive constant. The
+        condition demands that the function at the left end point be not much
+        bigger than the starting point (i.e. alpha = 0). This is an easy to
+        satisfy condition because by assumption, we are in a direction where
+        the function value is decreasing. The second and third conditions
+        together demand that there is at least one zero of the derivative in
+        between a and b. In addition to the interval, the update algorithm
+        requires a third point to be supplied. Usually, this point would lie
+        within the interval [a, b]. If the point is outside this interval,
+        the current interval is returned. If the point lies within the
+        interval, the behaviour of the function and derivative value at this
+        point is used to squeeze the original interval in a manner that
+        preserves the opposite slope conditions.
```

Comment:
Not a suggested change but I really like these explanation paragraphs! Helped to to understand it.